### PR TITLE
New `ProdMisfit` plugin

### LINKED ISSUE #754: Production misfit plugin
Create a plugin to plot production data misfit and coverage at specified dates:
- Prod misfit quantification and ranking of realizations in an ensemble
- Coverage plot of wells in an ensemble. 
- 
Read data from ecl/flow summary files (or optionally from pre-processed csv file)
Use well cummulative oil/water/gas prod data 

Example of misfit/ranking plot:
image

Example of coverage plot (1):
image

Example of coverage plot (2):
image

### PR DESCRIPTION
**Pull request by rnyb**
Work In Progress

---

### Contributor checklist

- [x] :tada: This PR closes #754.
- [x] :scroll: I have broken down my PR into the following tasks:
   - [x] Use EnsembleSummaryProvider and summary files on arrow format
   - [x] Visualization of prod misfit at selected time.
   - [x] Visualization of prod coverage at selected time.
   - [x] Heatmap representation of ensemble mean misfit for selected dates.
- [ ] :robot: I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR.
- [x] :book: I have considered adding a new entry in `CHANGELOG.md`, and added it if should be communicated there.

### CODE REVIEW

**Review by lindjoha — CHANGES_REQUESTED** [review]
I have started on a review, but will do more later

File: webviz_subsurface/plugins/_prod_misfit/_plugin.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        #         None,
+        #         None,
+        #     )
+        #     self.df_hist_mean[ens_name] = df_hist.groupby("DATE", as_index=False).mean()
+        #     self.df_hist_mean[ens_name].drop(columns=["REAL"], inplace=True)
+        #     # print(ens_name, "df_hist_mean:\n", self.df_hist_mean[ens_name])
+
+        self.set_callbacks(app)
+
+        logging.debug(f"Init done. Cummulative time: {time.time() - start}")
```
[the remark this answers is not in the document]
Comment (rnyb) [reply webviz_subsurface/plugins/_prod_misfit/_plugin.py:147]: will do

File: webviz_subsurface/plugins/_prod_misfit/_plugin.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+
+# --------------------------------
+def _check_well_collections(
+    well_collections: Optional[Dict[str, List[str]]], wells: dict
+) -> Dict[str, List[str]]:
+    """Check well collections vs well lists.
+    Any well not included in well collections is returned as Undefined."""
+
+    all_wells = []
```
[the remark this answers is not in the document]
Comment (rnyb) [reply webviz_subsurface/plugins/_prod_misfit/_plugin.py:188]: The  `_check_well_collections` function is an older version of `_get_well_collections` function and not used anymore, so i can remove it altogether.

File: webviz_subsurface/plugins/_prod_misfit/_plugin.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+
+# --------------------------------
+def _get_well_collections(
+    well_groups_file: Optional[str], wells: dict
+) -> Dict[str, List[str]]:
+    """Read csv file and create well_collections dictionary. Then check well collections
+    vs well lists. Any well not included in well collections is returned as Undefined."""
+
+    all_wells = []
```
[the remark this answers is not in the document]
Comment (rnyb) [reply webviz_subsurface/plugins/_prod_misfit/_plugin.py:222]: `all_wells = sorted(list(wells.values()))` will generate a list of lists so not sure if that would work as intended?

File: webviz_subsurface/plugins/_prod_misfit/_plugin.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+    for ens_wells in wells.values():
+        all_wells.extend(ens_wells)
+    all_wells = list(sorted(set(all_wells)))
+
+    if well_collections is None:
+        well_collections = {}
+        well_collections["Undefined"] = all_wells
+    else:
+        undefined_wells = []
+        all_collection_wells = []
```
[the remark this answers is not in the document]
Comment (rnyb) [reply webviz_subsurface/plugins/_prod_misfit/_plugin.py:198]: If i understand it correctly I think this won't work (TypeError: unhashable type: 'list')

File: webviz_subsurface/plugins/_prod_misfit/_callbacks.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+                relative_diff=relative_diff,
+            )
+        )
+
+        # dframe = self.df_diff_stat.copy()
+
+        # # --- apply date filter
+        # dframe = dframe.loc[dframe["DATE"].isin(dates)]
+
+        # # --- apply ensemble filter
```
Comment (lindjoha) [inline webviz_subsurface/plugins/_prod_misfit/_callbacks.py:307]: Remove old code?

File: webviz_subsurface/plugins/_prod_misfit/_callbacks.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    well_collections: Dict[str, List[str]],
+    selected_collection_names: list,
+    selected_wells: list,
+    combine_type: str = "intersection",
+) -> List[str]:
+    """Return union or intersection of well list and well collection lists."""
+
+    selected_collection_wells = []
+    for collection_name in selected_collection_names:
+        selected_collection_wells.extend(well_collections[collection_name])
```
Comment (lindjoha) [inline webviz_subsurface/plugins/_prod_misfit/_callbacks.py:339]: maybe like this? 
`selected_collection_wells = list(set([well_collections[name] for name in selected_collection_names])) `

**Review by lindjoha — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -10,15 +10,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 
 - [#981](https://github.com/equinor/webviz-subsurface/pull/981) - `WellAnalysis` - New plugin well analysis. One tab for well production overview plots and one for well control mode and network analysis.
-- [#987](https://github.com/equinor/webviz-subsurface/pull/987) - `SwatinitQC` - New plugin for QC of Water Initialization in Eclipse runs where the `SWATINIT` keyword has been used. 
+- [#987](https://github.com/equinor/webviz-subsurface/pull/987) - `SwatinitQC` - New plugin for QC of Water Initialization in Eclipse runs where the `SWATINIT` keyword has been used.
 - [#988](https://github.com/equinor/webviz-subsurface/pull/988) - `ParameterAnalysis` and `PropertyStatistics` - Switched to the VectorSelector component, and other various improvements.
```
Comment (anders-kiaer) [inline CHANGELOG.md:14]: Add the good news about prod misfit added to changelog?

File: webviz_subsurface/plugins/__init__.py [hunk 8]
```diff
@@ -36,6 +36,9 @@
 from ._parameter_distribution import ParameterDistribution
 from ._parameter_parallel_coordinates import ParameterParallelCoordinates
 from ._parameter_response_correlation import ParameterResponseCorrelation
+
+# from ._prod_misfit._plugin import ProdMisfit
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/__init__.py:40]:
```suggestion
```
Remove?

File: webviz_subsurface/plugins/_prod_misfit/_callbacks.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+import dash
+import webviz_core_components as wcc
+from dash.dependencies import Input, Output
+
+from .._simulation_time_series.types.provider_set import ProviderSet
+from ._layout import LayoutElements
+from .utils import make_dataframes as makedf
+from .utils import make_figures as makefigs
+
+# from ._property_serialization import GraphFigureBuilder
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_prod_misfit/_callbacks.py:12]: Remove?
```suggestion
```

File: webviz_subsurface/plugins/_prod_misfit/_plugin.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+        for collection_wells in well_collections.values():
+            all_collection_wells.extend(collection_wells)
+        all_collection_wells = list(set(all_collection_wells))
+        undefined_wells = [
+            well for well in all_wells if well not in all_collection_wells
+        ]
+        if len(undefined_wells) > 0:
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_prod_misfit/_plugin.py:191]:
```suggestion
        if undefined_wells:
```

File: webviz_subsurface/plugins/_prod_misfit/_plugin.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+                drop_list.append(well)
+                continue
+        if well not in wells:
+            wells.append(well)
+        if vector not in vectors:
+            vectors.append(vector)
+
+    wells, vectors = sorted(wells), sorted(vectors)
+
+    if len(vectors) == 0:
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_prod_misfit/_plugin.py:242]:
```suggestion
    if not vectors:
```

File: webviz_subsurface/plugins/_prod_misfit/_plugin.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        RuntimeError("No WOPT, WWPT or WGPT vectors found.")
+
+    if len(drop_list) > 0:
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_prod_misfit/_plugin.py:245]:
```suggestion
    if drop_list:
```

**Review by anders-kiaer — APPROVED** [review]

**Pull request merged**